Index: procps_commit/src/ps/common.h
===================================================================
--- procps_commit.orig/src/ps/common.h
+++ procps_commit/src/ps/common.h
@@ -329,6 +329,11 @@ makEXT(noop)
 #define FM_P 0x0010 /* -P */
 #define FM_M 0x0020 /* -M */
 //#define FM_T 0x0040 /* -T */
+#define FM_UsecUser 0x0040 /* --usec-user */
+#define FM_UsecProc 0x0080 /* --usec-proc */
+
+
+
 #define FM_F 0x0080 /* -F */  /* -F also sets the regular -f flags */
 
 /* sorting & formatting */
Index: procps_commit/src/ps/output.c
===================================================================
--- procps_commit.orig/src/ps/output.c
+++ procps_commit/src/ps/output.c
@@ -1537,6 +1537,71 @@ setREL1(ID_TGID)
   return len;
 }
 
+static int pr_usecuser(char *restrict const outbuf, const proc_t *restrict const pp){
+    size_t len;
+    unsigned int usec_user = 0xffffffff;
+    char filename[52];
+    ssize_t num_read;
+    int fd;
+    setREL1(ID_PID)
+
+    snprintf(filename, sizeof filename, "/proc/%d/attr/usec/ubits", rSv(ID_PID, s_int, pp));
+
+    fd = open(filename, O_RDONLY);
+    if (fd == -1) {
+        outbuf[0] = '-';
+        outbuf[1] = '\0';
+        len = 1;
+        return len;
+    }
+
+    ssize_t bytes_written = read(fd, &usec_user, sizeof(usec_user));
+    if (bytes_written == -1) {
+        close(fd);
+        outbuf[0] = '-';
+        outbuf[1] = '\0';
+        len = 1;
+        return len;
+    }
+    close(fd);
+
+    unsigned int first_8_bits = (usec_user >> 24);
+    snprintf(outbuf, 5, "0x%02X", first_8_bits);
+    outbuf[5] = '\0';
+    return 4;
+}
+
+static int pr_usecproc(char *restrict const outbuf, const proc_t *restrict const pp){
+    size_t len;
+    char filename[52];
+    ssize_t num_read;
+    int fd;
+    setREL1(ID_PID)
+
+    snprintf(filename, sizeof filename, "/proc/%d/attr/usid", rSv(ID_PID, s_int, pp));
+
+    fd = open(filename, O_RDONLY);
+    if (fd == -1) {
+        outbuf[0] = '-';
+        outbuf[1] = '\0';
+        len = 1;
+        return len;
+    }
+
+    ssize_t bytes_read = read(fd, outbuf, OUTBUF_SIZE-1);
+    if (bytes_read == -1) {
+        close(fd);
+        outbuf[0] = '-';
+        outbuf[1] = '\0';
+        len = 1;
+        return len;
+    }
+    close(fd);
+    outbuf[bytes_read] = '\0';
+
+    return bytes_read;
+}
+
 /************************ Linux miscellaneous ***************************/
 static int pr_agid(char *restrict const outbuf, const proc_t *restrict const pp){
 setREL1(AUTOGRP_ID)
@@ -1912,6 +1977,8 @@ static const format_struct format_array[
 {"unit",      "UNIT",    pr_sd_unit,       PIDS_SD_UNIT,            31,    LNX,  ET|LEFT},
 {"upr",       "UPR",     pr_nop,           PIDS_noop,                3,    BSD,  TO|RIGHT}, /*usrpri*/
 {"uprocp",    "UPROCP",  pr_nop,           PIDS_noop,                8,    BSD,  AN|RIGHT},
+{"usec-proc", "USEC-PROC",pr_usecproc,     PIDS_ID_TGID,            31,    SGI,  ET|LEFT},
+{"usec-user", "USEC-USER",pr_usecuser,     PIDS_ID_TGID,            31,    SGI,  ET|LEFT},
 {"user",      "USER",    pr_euser,         PIDS_ID_EUSER,            8,    U98,  ET|USER},  /* BSD n forces this to UID */
 {"userns",    "USERNS",  pr_userns,        PIDS_NS_USER,            10,    LNX,  ET|RIGHT},
 {"usertime",  "USER",    pr_nop,           PIDS_noop,                4,    DEC,  ET|RIGHT},
@@ -1935,7 +2002,7 @@ static const format_struct format_array[
 {"wname",     "WCHAN",   pr_wchan,         PIDS_WCHAN_NAME,          6,    SGI,  TO|WCHAN}, /* opposite of nwchan */
 {"wops",      "WOPS",    pr_wops,          PIDS_IO_WRITE_OPS,        5,    LNX,  TO|RIGHT},
 {"xstat",     "XSTAT",   pr_nop,           PIDS_noop,                5,    BSD,  AN|RIGHT},
-{"zone",      "ZONE",    pr_context,       PIDS_ID_TGID,            31,    SUN,  ET|LEFT},  // Solaris zone == Linux context?
+{"zone",      "ZONE",    pr_context,       PIDS_ID_PID,            31,    SUN,  ET|LEFT},  // Solaris zone == Linux context?
 {"zoneid",    "ZONEID",  pr_nop,           PIDS_noop,               31,    SUN,  ET|RIGHT}, // Linux only offers context names
 {"~",         "-",       pr_nop,           PIDS_noop,                1,    LNX,  AN|RIGHT}  /* NULL would ruin alphabetical order */
 };
Index: procps_commit/src/ps/parser.c
===================================================================
--- procps_commit.orig/src/ps/parser.c
+++ procps_commit/src/ps/parser.c
@@ -836,6 +836,8 @@ static const char *parse_gnu_option(void
   {"signames",      &&case_signames},
   {"sort",          &&case_sort},
   {"tty",           &&case_tty},
+  {"usec-proc",     &&case_usec_proc},
+  {"usec-user",     &&case_usec_user},
   {"user",          &&case_user},        /* euid */
   {"version",       &&case_version},
   {"width",         &&case_width},
@@ -1041,6 +1043,14 @@ static const char *parse_gnu_option(void
     trace("--context\n");
     format_flags |= FF_Fc;
     return NULL;
+  case_usec_user:
+    trace("--usec-user\n");
+    format_modifiers |= FM_UsecUser;
+    return NULL;
+  case_usec_proc:
+    trace("--usec-proc\n");
+    format_modifiers |= FM_UsecProc;
+    return NULL;
 }
 
 /*************** process trailing PIDs  **********************/
Index: procps_commit/src/ps/sortformat.c
===================================================================
--- procps_commit.orig/src/ps/sortformat.c
+++ procps_commit/src/ps/sortformat.c
@@ -752,6 +752,12 @@ static const char *generate_sysv_list(vo
   if(format_modifiers & FM_M){
     PUSH("label");  /* Mandatory Access Control */
   }
+  if(format_modifiers & FM_UsecUser){
+    PUSH("usec-user");  /* Mandatory Access Control */
+  }
+  if(format_modifiers & FM_UsecProc){
+    PUSH("usec-proc");  /* Mandatory Access Control */
+  }
   return NULL;
 }
 
@@ -943,6 +949,16 @@ did_lwp:
       fn->next=format_list;
       format_list=fn;
     }
+    if(format_modifiers & FM_UsecUser){
+        fn = do_one_spec("usec-user", NULL);
+        fn->next=format_list;
+        format_list=fn;
+    }
+    if(format_modifiers & FM_UsecProc){
+        fn = do_one_spec("usec-proc", NULL);
+        fn->next=format_list;
+        format_list=fn;
+    }
     /* Do personality-specific translations not covered by format_flags.
      * Generally, these only get hit when personality overrides unix output.
      * That (mostly?) means the Digital and Debian personalities.
